[quant] Support lowering for quantized embedding byte operator

jerryzh168 opened the pull request:
Summary: This PR adds lowering for embedding in quantization in executorch flow

Test Plan: buck run executorch/exir/tests:quant_fusion_pass -- "executorch.exir.tests.test_quant_fusion_pass.TestQuantFusionPass.test_embedding_byte"

Reviewed By: qihqi

Differential Revision: D41673139

vkuzo commented on torch/ao/nn/quantized/reference/modules/rnn.py:
```diff
@@ -294,10 +301,14 @@ def __init__(self, mode: str, input_size: int, hidden_size: int,
             for wn in self._flat_weights_names:
                 if wn.startswith("weight"):
                     weight_qparams_dict[wn] = weight_qparams
+            weight_qparams_dict["is_decompoed"] = False  # type: ignore[assignment]
```
nit: remove?

vkuzo replied:
or, fix typo.  Not sure what the intent was.

jerryzh168 replied:
probably a mistake, I think I updated L300 to include this

vkuzo commented on torch/ao/nn/quantized/reference/modules/utils.py:
```diff
@@ -50,6 +50,10 @@ def _init_weight_qparams(self, weight_qparams, device):
             self.register_buffer(
                 "weight_axis", torch.tensor(0, dtype=torch.int, device=device))
+        self.is_decomposed: bool = weight_qparams["is_decomposed"]
+        # store weight_axis as weight_axis_int due to some constraints of torchdynamo.expport
```
nit: typo in export

vkuzo commented on torch/ao/nn/quantized/reference/modules/utils.py:
```diff
@@ -98,13 +120,109 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
+        weight_dtype: torch.dtype,
+        weight_scale: torch.Tensor,
+        weight_zero_point: torch.Tensor,
+        weight_axis: int):
```
can we add a return type hint

vkuzo commented on torch/ao/nn/quantized/reference/modules/utils.py:
```diff
@@ -98,13 +120,109 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
+        weight_scale: torch.Tensor,
+        weight_zero_point: torch.Tensor,
+        weight_axis: int):
+    # TODO: get the quant_min and quant_max from activation_post_process
```
is this for before land or for later?

jerryzh168 replied:
this is for later, we don't have this for all observers right now I think

jerryzh168 replied:
I think we should probably have a BE task to make quant_min/quant_max a required field

vkuzo commented on torch/ao/nn/quantized/reference/modules/utils.py:
```diff
@@ -98,13 +120,109 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
+        weight_zero_point: torch.Tensor,
+        weight_axis: int):
+    # TODO: get the quant_min and quant_max from activation_post_process
+    _DTYPE_TO_QVALUE_BOUNDS = {
```
is this for before land or for later?

vkuzo commented on torch/ao/nn/quantized/reference/modules/utils.py:
```diff
@@ -98,13 +120,109 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
+        torch.qint32: torch.int32,
+    }
+    if weight_qscheme == torch.per_tensor_affine:
+        if weight_dtype in [torch.quint8, torch.qint8, torch.qint32]:
+            weight_dtype_ = _QDTYPE_TO_DTYPE[weight_dtype]
```
nit: the current naming is confusing, how about something like

```
if weight_dtype in [...]:
    weight_dtype_decomposed = _QDTYPE_TO_UNDERLYING_INT_REPR_DTYPE[...]
```

vkuzo commented on torch/ao/quantization/backend_config/executorch.py:
```diff
@@ -56,6 +56,13 @@
     is_dynamic=True,
 )
 
+executorch_weight_only_quint8_dtype_config = DTypeConfig(
```
would be great to have a TODO to rename this file, since there is no executorch backend.  Maybe it should be `executorch_with_qnnpack_backing`?

when other targets will start to be supported by executorch runtime (DSP chips, etc), they may have different capabilities compared to what is being added here

vkuzo reviewed (approved):
would be great to do some of the TODOs before landing, specifically the one which looks up qmin/qmax from observer

jerryzh168 committed:
[quant] Support lowering for quantized embedding byte operator (#91159)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/91159

This PR adds lowering for embedding in quantization in executorch flow

Test Plan: buck run executorch/exir/tests:quant_fusion_pass -- "executorch.exir.tests.test_quant_fusion_pass.TestQuantFusionPass.test_embedding_byte"

Reviewed By: vkuzo, qihqi

Differential Revision: D41673139

fbshipit-source-id: 1447bf6e8b91e984cf973ea9f1adf80d92554382
